refactor: Slash commands (#936)
discordjs/guide: guide/interactions/replying-to-slash-commands.md

# Replying to slash commands

::: tip
This page is a follow-up to [Registering slash commands](./registering-slash-commands.md) and assumes you have the slash commands that are explained in that page already registered.
:::

Here's a flowchart to help you understand visually *which* methods are you allowed to use at any given time. Feel free to refer back to this flowchart as you read the rest of this page.

![](./images/flowchart.png)

## Receiving interactions

> A slash command is just one type of [interaction](https://discord.com/developers/docs/interactions/application-commands#Slash-commands). Therefore, you can make your bot listen to the [`interactionCreate`](https://discord.js.org/#/docs/main/stable/class/Client?scrollTo=e-interactionCreate) event to receive incoming interactions.

Reviewer: ```suggestion
A slash command is just one type of [interaction](https://discord.com/developers/docs/interactions/application-commands#slash-commands). Therefore, you can make your bot listen to the [`interactionCreate`](https://discord.js.org/#/docs/main/stable/class/Client?scrollTo=e-interactionCreate) event to receive incoming interactions.
```

  },
  replied: false,
  ephemeral: null,
  webhook: InteractionWebhook { id: '123' }
}
```

If you did, then your bot is receiving incoming interactions properly!

## Reply

::: warning
You must reply or defer the interaction within **three seconds** of receiving it.
:::

Replying to an interaction is really simple. We can just call the `.reply()` method.

:::: code-group
::: code-group-item index.js
```js {11-21}
const { Client, Intents } = require('discord.js');
const { token } = require('./config.json');

const client = new Client({ intents: [Intents.FLAGS.GUILDS] });

client.once('ready', () => {
	console.log('Ready!');
});

client.on('interactionCreate', async interaction => {
	if (!interaction.isCommand()) return;

	const { commandName } = interaction;

	if (commandName === 'ping') {
		await interaction.reply('Pong!');
	} else if (commandName === 'server') {
		await interaction.reply(`Server name: ${interaction.guild.name}\nTotal members: ${interaction.guild.memberCount}`);
	} else if (commandName === 'user') {
		await interaction.reply(`Your tag: ${interaction.user.tag}\nYour id: ${interaction.user.id}`);
	}
});

client.login(token);
```
:::
::::

::: tip

> The `index.js` file we have here doesn't have a command handler (for the sake of simplicity during explanation), but the concepts and examples that are explained in this page are also applicable to the command handler version.

Reviewer: ```suggestion
The `index.js` file you have here doesn't have a command handler (for the sake of simplicity during explanation), but the concepts and examples that are explained in this page are also applicable to the command handler version.
```

:::

A slash command is just one type of interaction. Therefore, you should first check if the interaction is a slash command by calling `interaction.isCommand()` (line 11).

Next, you should check the name of the command by accessing the `commandName` property (line 13). Once you have the name of the command, you can use some `if`/`else if` statements to reply accordingly. 

Finally, you reply to the interaction by calling `interaction.reply()`.

Let's start the bot:
```sh:no-line-numbers
node index.js
```

> and then use the `/ping` command in your guild. Your bot should've replied and you should see something like this in your Discord:

Reviewer: ```suggestion
And then use the `/ping` command in your guild. Your bot should've replied and you should see something like this in your Discord:
```

```js:no-line-numbers {3}
await interaction.deferReply();
// do some task that takes a long time to finish
await interaction.editReply('The task has been completed!');
```

<DiscordMessages>
	<DiscordMessage profile="bot">
		<template #interactions>
			<DiscordInteraction profile="user" :command="true">ping</DiscordInteraction>
		</template>
		The task has been completed!
	</DiscordMessage>
</DiscordMessages>

::: tip
Technically speaking, you could call `.reply()` right away with a message like `"Processing. Please wait..."` and treat that as a deferral. It's up to you to choose between using `.reply()` or `.deferReply()` within the first three seconds of receiving the interaction.

The only technical difference between `.reply()` and `.deferReply()` is that `.deferReply()` will display a `<Bot name> is thinking...` message for 15 minutes and then turn into an "**Interaction failed**" error message once the 15 minutes window has expired. To avoid this error message, you should call `.editReply()` or `.followUp()` at least once within that 15 minutes window.
:::

## Edit

### Edit a reply

> After sending an initial reply or defer (which must be done within **three seconds** of receiving the interaction), you will have a **15 minutes** window to edit the reply or send followup messages however many times you want. We will look at followup messages in the next section.

Reviewer: ```suggestion
After sending an initial reply or defer (which must be done within **three seconds** of receiving the interaction), you will have a **15 minutes** window to edit the reply or send follow-up messages however many times you want. We will look at follow-up messages in the next section.
```

```js:no-line-numbers
await interaction.reply('Pong!');
```

<DiscordMessages>
	<DiscordMessage profile="bot">
		<template #interactions>
			<DiscordInteraction
				profile="user"
				:command="true"
			>ping</DiscordInteraction>
		</template>
		Pong!
	</DiscordMessage>
</DiscordMessages>

```js:no-line-numbers {2}
await interaction.reply('Pong!');
await interaction.editReply('Pong edited');
```

<DiscordMessages>
	<DiscordMessage profile="bot" :edited="true">
		<template #interactions>
			<DiscordInteraction
				profile="user"
				:command="true"
			>ping</DiscordInteraction>
		</template>
		Pong edited
	</DiscordMessage>
</DiscordMessages>

### Edit a deferReply

You can also call `.editReply()` **after** calling `.deferReply()`.

```js:no-line-numbers
await interaction.deferReply();
```

![](./images/deferreply.png)

```js:no-line-numbers {2}
await interaction.deferReply();
await interaction.editReply('The task has been completed!');
```

<DiscordMessages>
	<DiscordMessage profile="bot">
		<template #interactions>
			<DiscordInteraction profile="user" :command="true">ping</DiscordInteraction>
		</template>
		The task has been completed!
	</DiscordMessage>
</DiscordMessages>

> ## Followup

Reviewer: ```suggestion
## Follow-up
```

> ### Followup a reply

Reviewer: ```suggestion
### Follow-up a reply
```

> You can send additional messages after the initial reply. Just like edits, followups also have a **15 minutes** window.

Reviewer: ```suggestion
You can send additional messages after the initial reply. Just like edits, follow-ups also have a **15 minutes** window.
```

You can call `.followUp()` **after** calling `.reply()`.

```js:no-line-numbers
await interaction.reply('pong');
await interaction.followUp('pong again');
```

![](./images/followup.png)

> ### Followup a deferReply

Reviewer: ```suggestion
### Follow-up a deferReply
```

> You can also call `.followUp()` **after** calling `.deferReply()`. However, a followup to a deferral does **not** behave the same way as a followup to a reply. **The first followup will edit the deferral instead**. Additional messages will start from **the second followup**.

Reviewer: ```suggestion
You can also call `.followUp()` **after** calling `.deferReply()`. However, a follow-up to a deferral does **not** behave the same way as a followup to a reply. **The first follow-up will edit the deferral instead**. Additional messages will start from **the second follow-up**.
```

await interaction.editReply({ content: 'pong edited', ephemeral: true });
```

The reply will **not** change from non-ephemeral to ephemeral, even though you called `.editReply()` with `ephemeral: true`.

<DiscordMessages>
	<DiscordMessage profile="bot">
		<template #interactions>
			<DiscordInteraction
				profile="user"
				:command="true"
        :ephemeral="false"
			>ping</DiscordInteraction>
		</template>
		Pong!
	</DiscordMessage>
</DiscordMessages>
<br />
<DiscordMessages>
	<DiscordMessage profile="bot" :edited="true">
		<template #interactions>
			<DiscordInteraction
				profile="user"
				:command="true"
			>ping</DiscordInteraction>
		</template>
		Pong edited
	</DiscordMessage>
</DiscordMessages>

---

The same applies to `.deferReply()` into `.editReply()`. 

The following **will stay non-ephemeral**.

```js:no-line-numbers
await interaction.deferReply({ ephemeral: false });
await interaction.editReply({ content: 'pong edited', ephemeral: true });
```

And the following **will stay ephemeral**.

```js:no-line-numbers
await interaction.deferReply({ ephemeral: true })
await interaction.editReply({ content: 'pong edited', ephemeral: false })
```

> ### Ephemeral followup

Reviewer: ```suggestion
### Ephemeral follow-up
```

> The ephemeral state of followups are independent of the initial reply. This is because the followup is a separate message.

Reviewer: ```suggestion
The ephemeral state of followups are independent of the initial reply. This is because the follow-up is a separate message.
```

```js:no-line-numbers
await interaction.reply({ content: 'pong', ephemeral: false })
await interaction.followUp({ content: 'first followup', ephemeral: true })
await interaction.followUp({ content: 'second followup', ephemeral: true })
```

![](./images/followup-ephemeral.png)

> As you saw in the previous section about [followup a deferReply](#followup-a-deferreply), the first followup to a deferral will replace the deferral instead of sending a new message. As a result, this will also affect the ephemeral state of the first followup:

Reviewer: ```suggestion
As you saw in the previous section about [follow-up a deferReply](#followup-a-deferreply), the first follow-up to a deferral will replace the deferral instead of sending a new message. As a result, this will also affect the ephemeral state of the first follow-up:
```

await message.awaitMessageComponent({ filter, componentType: 'BUTTON' });
```

Refer to the [Buttons](./buttons.md) page for more information about how to build one.

Another way to fetch the reply is to provide a `fetchReply: true` to the `.reply()` method.

```js:no-line-numbers
const message = await interaction.reply({ content: 'Pong!', fetchReply: true });
```

## Deleting the reply after sending it

::: danger
You **cannot** delete an ephemeral message.
:::

You can also delete the reply after sending it.

```js:no-line-numbers
await interaction.reply('Pong!');
await interaction.deleteReply();
```

## Options

In this section, we'll cover how to access the values of a command's options.

Let's take the `/ping` command from [Registering slash commands](./registering-slash-commands.md) as example. If a user used your `/ping` command like this:

![](./images/option-value.png)

Then you can retrieve the value "`hello this is some string input`" from the `option-name` option as shown below:

```js:no-line-numbers {1-2}
const userInput = interaction.options.getString('option-name');
console.log(userInput); // expected output: "hello this is some string input"
```

> We can access a command's options with the `.options` property, which is an instance of [`CommandInteractionOptionResolver`](https://discord.js.org/#/docs/main/stable/class/CommandInteractionOptionResolver). 

Reviewer: ```suggestion
You can access a command's options with the `.options` property, which is an instance of [`CommandInteractionOptionResolver`](https://discord.js.org/#/docs/main/stable/class/CommandInteractionOptionResolver). 
```

> Then, we call `.getString('option-name')` to retrieve the value that the user provided to the `option-name` option.

Reviewer: ```suggestion
Then, you call `.getString('option-name')` to retrieve the value that the user provided to the `option-name` option.
```

`.getNumber()` | [`Number`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)
`.getMember()` | [`GuildMember`](https://discord.js.org/#/docs/main/stable/class/GuildMember)

Example:
```js:no-line-numbers
const userInput = interaction.options.getString('option-name');
```

### Predefined choices

Let's suppose you have a `ping` command with the following structure:

```js:no-line-numbers {8-10}
new SlashCommandBuilder()
  .setName('ping')
  .setDescription('Replies with pong!')
  .addStringOption(option => 
    option.setName('option-name')
      .setDescription('some description')
      .setRequired(true)
      .addChoice('name1', 'value1')
      .addChoice('name2', 'value2')
      .addChoice('name2', 'value3'))
```

The first parameter of `.addChoice()` is the name of the choice that will be displayed in Discord, while the second parameter is the value of that choice that your bot will receive.

If a user used your `/ping` command and selected the `name2` choice in the `option-name` option, as shown below:

![](./images/choices-selected.png)

> then the value that your bot receives is the string `"value2"`.

Reviewer: ```suggestion
Then the value that your bot receives is the string `"value2"`.
```
